driver-core/src/main/com/mongodb/client/model/expressions/DocumentExpression.java

```diff
@@ -25,86 +28,500 @@
+        Assertions.notNull("fieldName", fieldName);
+        Assertions.notNull("other", other);
         return getMap(fieldName, ofMap(other));
     }
 
-    <T extends Expression> ArrayExpression<T> getArray(String fieldName);
+    /**
+     * Returns the {@linkplain ArrayExpression array} value of the field
+     * with the provided {@code fieldName}.
+     *
+     * <p>Warning: The type and presence of the resulting value is not
+     * enforced by the API. The use of this method is an
+     * unchecked assertion that the document
+     * {@linkplain #has(String) has} the named field,
+     * and the field value is of the specified raw type,
+     * and the field value's type has the specified type argument.
+     *
+     * @mongodb.server.release 5.0
+     * @param fieldName the name of the field.
+     * @return the resulting value.
+     * @param <T> the type.
+     */
+    @MqlUnchecked({PRESENT, TYPE})
+    <T extends Expression> ArrayExpression<@MqlUnchecked(TYPE_ARGUMENT) T> getArray(String fieldName);
 
-    <T extends Expression> ArrayExpression<T> getArray(String fieldName, ArrayExpression<? extends T> other);
+    /**
+     * Returns the {@linkplain ArrayExpression array} value of the field
+     * with the provided {@code fieldName},
+     * or the {@code other} value if the field is not an array
+     * or if the document {@linkplain #has} no such field.
+     *
+     * <p>Warning: The type argument of the resulting value is not
+     * enforced by the API. The use of this method is an
+     * unchecked assertion that the type argument is correct.
+     *
+     * @mongodb.server.release 5.0
+     * @param fieldName the name of the field.
+     * @param other the other value.
+     * @return the resulting value.
+     * @param <T> the type.
+     */
+    <T extends Expression> ArrayExpression<@MqlUnchecked(TYPE_ARGUMENT) T> getArray(String fieldName, ArrayExpression<? extends T> other);
 
+    /**
+     * Returns a document with the same fields as {@code this} document, but
+     * with any fields present in the {@code other} document overwritten with
+     * the fields of that other document. That is, fields from both this and the
+     * other document are merged, with the other document having priority.
+     *
+     * <p>This does not affect the original document.
+     *
+     * @param other the other document.
+     * @return the resulting value.
+     */
     DocumentExpression merge(DocumentExpression other);
 
-    MapExpression<Expression> asMap();
+    /**
+     * {@code this} document as a {@linkplain MapExpression map}.
+     *
+     * <p>Warning: The type argument of the resulting value is not
+     * enforced by the API. The use of this method is an
+     * unchecked assertion that the type argument is correct.
+     *
+     * @return the resulting value.
+     * @param <T> the type.
+     */
+
+    <T extends Expression> MapExpression<T> asMap();
```

Comment:
In https://github.com/mongodb/mongo-java-driver/pull/1059 this method returns `MapExpression<Expression>`, which is why the table does not list it. But now that it has changed, it also needs to be annotated.

Reply:
Moved to next PR due to sensitive rebase.